modules/tinymce/src/core/main/ts/api/html/SaxParser.ts

```diff
@@ -119,38 +119,55 @@ const isInvalidUri = (settings: SaxParserSettings, uri: string, tagName: string)
 };
 
 /**
- * Returns the index of the end tag for a specific start tag. This can be
- * used to skip all children of a parent element from being processed.
+ * Returns the index of the matching end tag for a specific start tag. This can
+ * be used to skip all children of a parent element from being processed.
  *
  * @private
- * @method findEndTagIndex
+ * @method findMatchingEndTagIndex
  * @param {tinymce.html.Schema} schema Schema instance to use to match short ended elements.
  * @param {String} html HTML string to find the end tag in.
  * @param {Number} startIndex Index to start searching at should be after the start tag.
  * @return {Number} Index of the end tag.
  */
-const findEndTagIndex = (schema: Schema, html: string, startIndex: number): number => {
-  let count = 1, index, matches;
-
+const findMatchingEndTagIndex = (schema: Schema, html: string, startIndex: number): number => {
+  const startTagRegExp = /<([!?\/])?([A-Za-z0-9\-_:.]+)/g;
+  const endTagRegExp = /(?:\s(?:[^'">]+(?:"[^"]*"|'[^']*'))*[^"'>]*(?:"[^">]*|'[^'>]*)?|\s*|\/)>/g;
   const shortEndedElements = schema.getShortEndedElements();
-  const tokenRegExp = /<([!?\/])?([A-Za-z0-9\-_:.]+)(\s(?:[^'">]+(?:"[^"]*"|'[^']*'))*[^"'>]*(?:"[^">]*|'[^'>]*)?|\s*|\/)>/g;
-  tokenRegExp.lastIndex = index = startIndex;
-
-  while ((matches = tokenRegExp.exec(html))) {
-    index = tokenRegExp.lastIndex;
-
-    if (matches[1] === '/') { // End element
-      count--;
-    } else if (!matches[1]) { // Start element
-      if (matches[2] in shortEndedElements) {
-        continue;
+  let count = 1, index = startIndex;
+
+  // keep finding HTML tags (opening, closing, or neither like comments or <br>s
+  while (count !== 0) {
+    startTagRegExp.lastIndex = index;
+
+    // ideally, we only want to run through this the once - but sometimes the startTagRegExp will give us false positives (things that begin
+    // like tags, but don't end like them) and so we might need to bump up its lastIndex and try again.
+    while (true) {
+      const startMatch = startTagRegExp.exec(html);
+      if (startMatch === null) {
+        // doesn't matter what count is, we've run out of HTML tags
+        return index;
       }
+      if (startMatch[1] === '!') {
```

Comment:
Else if here please, as this is a continuation of the branching, not something new.

Reply:
Also, do we care that this'll incorrectly run for `<!DOCTYPE`? We shouldn't ever run into that when trying to find an end tag though.

Reply:
Another issue here is what we discussed yesterday whereby this won't be supporting CDATA. Please do what we said there and log a Jira and add a TODO comment here.